lib/StackAnalysis/StackAnalysis.cpp

```diff
@@ -441,6 +2060,167 @@ bool StackAnalysis::runOnModule(Module &M) {
               getName(Function.Entry) << (Function.Force ? " (forced)" : ""));
   }
 
+  using BasicBlockToNodeMapTy = llvm::DenseMap<BasicBlock *, BasicBlockNode *>;
+  BasicBlockToNodeMapTy BasicBlockNodeMap;
+
+  // Queue to be populated with the CFEP
+  llvm::SmallVector<BasicBlock *, 8> Worklist;
+  SmallCallGraph CG;
+
+  // Create over-approximated call graph
+  for (CFEP &Function : Functions) {
+    BasicBlockNode Node{ Function.Entry };
+    BasicBlockNode *GraphNode = CG.addNode(Node);
+    BasicBlockNodeMap[Function.Entry] = GraphNode;
+  }
+
+  for (CFEP &Function : Functions) {
+    llvm::SmallSet<BasicBlock *, 8> Visited;
+    BasicBlockNode *StartNode = BasicBlockNodeMap[Function.Entry];
+    revng_assert(StartNode != nullptr);
+    Worklist.emplace_back(Function.Entry);
+
+    while (!Worklist.empty()) {
+      BasicBlock *Current = Worklist.pop_back_val();
+      Visited.insert(Current);
+
+      if (isFunctionCall(Current)) {
+        // If not an indirect call, add the node to the CG
+        if (BasicBlock *Callee = getFunctionCallCallee(Current)) {
+          auto *Node = BasicBlockNodeMap[Callee];
+          StartNode->addSuccessor(Node);
+        }
+        BasicBlock *Next = getFallthrough(Current);
+        if (!Visited.count(Next))
+          Worklist.push_back(Next);
+      }
+
+      for (BasicBlock *Successor : successors(Current)) {
+        if (!GCBI::isPartOfRootDispatcher(Successor)
+            && !Visited.count(Successor))
+          Worklist.push_back(Successor);
+      }
+    }
+  }
+
+  // Collect entry blocks so that their incoming edge will be a root empty node
+  llvm::SmallVector<BasicBlockNode *, 4> EntryBlocks;
+  for (auto &SCC : llvm::make_range(llvm::scc_begin(&CG), llvm::scc_end(&CG))) {
+    llvm::DenseSet<BasicBlock *> Predecessors;
+
+    // Loop over the elements of the vector containing the nodes creating an SCC
+    for (const auto &Node : SCC) {
+      // Enumerate the predecessors of a node
+      for (const auto &Pred : llvm::predecessors(Node->BB))
+        Predecessors.insert(Pred);
+    }
+
+    for (const auto &Node : SCC)
+      Predecessors.erase(Node->BB);
+
+    // Is the set containing the predecessors of each node equal to the set of
+    // the nodes themselves? If so, Predecessors is empty.
+    if (Predecessors.empty()) {
+      // Elect one of the elements as candidate entry point. `SCC.front` points
+      // to the first element of each SCC subgraph.
+      EntryBlocks.emplace_back(SCC.front());
+    }
+  }
+
+  BasicBlockNode FakeNode{ nullptr };
+  BasicBlockNode *RootNode = CG.addNode(FakeNode);
+  CG.setEntryNode(RootNode);
+  for (const auto &Node : EntryBlocks)
+    RootNode->addSuccessor(Node);
+
+  std::unique_ptr<raw_fd_ostream> OutputCG;
+  if (CallGraphOutputPath.getNumOccurrences() == 1) {
+    if (std::filesystem::exists(CallGraphOutputPath.c_str())) {
```

Comment:
Please don't use `std::filesystem` for now: it's not supported on Windows by libcxx.